ethane-types/src/ethereum_type.rs

```diff
@@ -0,0 +1,414 @@
+use serde::de::Visitor;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::convert::{From, TryFrom, TryInto};
+
+use crate::be_bytes::BeBytes;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct EthereumType<const N: usize, const H: bool>([u8; N]);
+
+impl<const N: usize, const H: bool> Serialize for EthereumType<N, H> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&self.to_string())
+    }
+}
+
+impl<'de, const N: usize, const H: bool> Deserialize<'de> for EthereumType<N, H> {
+    fn deserialize<D>(deserializer: D) -> Result<EthereumType<N, H>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_str(EthereumTypeVisitor::<N, H>)
+    }
+}
+
+struct EthereumTypeVisitor<const N: usize, const H: bool>;
+
+impl<'de, const N: usize, const H: bool> Visitor<'de> for EthereumTypeVisitor<N, H> {
+    type Value = EthereumType<N, H>;
+
+    fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "a hex string")
+    }
+
+    fn visit_str<T: serde::de::Error>(self, value: &str) -> Result<Self::Value, T> {
+        let result = Self::Value::try_from(value)
+            .map_err(|e| T::custom(format!("Deserialization error: {:?}", e)))?;
+        Ok(result)
+    }
+
+    fn visit_string<T: serde::de::Error>(self, value: String) -> Result<Self::Value, T> {
+        let result = Self::Value::try_from(value.as_str())
+            .map_err(|e| T::custom(format!("Deserialization error: {:?}", e)))?;
+        Ok(result)
+    }
+}
+
+impl<const N: usize, const H: bool> EthereumType<N, H> {
+    /// Represents inner data as a byte slice.
+    #[inline]
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+
+    /// Consumes `self` to uncover the underlying fixed array.
+    #[inline]
+    pub fn into_bytes(self) -> [u8; N] {
+        self.0
+    }
+
+    /// Creates a zero instance of the type.
+    #[inline]
+    pub fn zero() -> Self {
+        Self([0_u8; N])
+    }
+
+    /// Tries to parse an integer type that implements the `BeBytes` trait.
+    ///
+    /// Checks whether the integer can be safely casted into the given type
+    /// and returns an error if it doesn't fit.
+    #[inline]
+    pub fn try_from_int<const L: usize>(value: impl BeBytes<L>) -> Result<Self, ConversionError> {
+        if N >= L {
+            let mut data = [0_u8; N];
+            data[N - L..].copy_from_slice(&value.be_bytes()[..]);
+            Ok(Self(data))
+        } else {
+            Err(ConversionError::TryFromIntError(format!(
+                "input does not fit into {} bytes",
+                N
+            )))
+        }
+    }
+
+    /// Parses an integer type without checking whether it can be safely casted
+    /// into the given type.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the input data doesn't fit into the type, i.e. `N < L`.
```

Comment:
Isn't it a prob;em in case of wasm? I guess in case of panic the caller side won't be notified about the panic, isn't it?

Reply:
Yeah, I guess, that's right, but this function was specifically written like this. The aim is to have an unchecked conversion function that doesn't return a `Result`, so it is faster and more convenient to use. This comes at the cost that the user is responsible to ensure the conversion never panics.

On the other hand, there is a `try_from_int` function, that returns a result which can be used alternatively.